src/libmongoc/tests/unified/operation.c

```diff
@@ -3202,18 +3203,279 @@ operation_assert_session_unpinned (test_t *test,
+                               bson_error_t *error)
+{
+   BSON_ASSERT_PARAM (em);
+   BSON_ASSERT (id || true);
+   BSON_ASSERT (error || true);
+
+   if (!id) {
+      // Nothing to do.
+      return true;
+   }
+
+   const entity_t *const entity = entity_map_get (em, id, error);
+
+   // If the entity does not exist, the test runner MUST create it with the
+   // type of BSON array.
+   if (!entity) {
+      return entity_map_add_bson_array (em, id, error);
+   }
+
+   // If the entity exists and is of type BSON array, the test runner MUST do
+   // nothing.
+   else if (strcmp (entity->type, "bson_array") == 0) {
+      return true;
+   }
+
+   // If the entity exists and is of a different type, the test runner MUST
+   // raise an error.
+   else {
+      test_set_error (
+         error, "loop entity %s exists but is not of type BSON array", id);
+      return false;
+   }
+}
+
+static bool
+create_loop_size_t_entity (entity_map_t *em,
+                           const char *id,
+                           bson_error_t *error)
+{
+   BSON_ASSERT_PARAM (em);
+   BSON_ASSERT (id || true);
+   BSON_ASSERT (error || true);
+
+   if (!id) {
+      return true;
+   }
+
+   const entity_t *const entity = entity_map_get (em, id, error);
+
+   // If the entity of the specified name already exists, the test runner MUST
+   // raise an error.
+   if (entity) {
+      test_set_error (
+         error, "loop entity %s already exists when it should not", id);
+      return false;
+   }
+
+   return entity_map_add_size_t (em, id, bson_malloc0 (sizeof (size_t)), error);
+}
+
+static void
+increment_loop_counter (entity_map_t *em, const char *id)
+{
+   BSON_ASSERT_PARAM (em);
+   BSON_ASSERT (id || true);
+
+   if (id) {
+      bson_error_t error = {0};
+      size_t *const counter = entity_map_get_size_t (em, id, &error);
+
+      // Entity should always be valid at this point.
+      ASSERT_OR_PRINT (counter, error);
+
+      *counter += 1u;
+   }
+}
+
+static bool
+append_loop_error (entity_map_t *em,
+                   const char *id,
+                   bson_error_t *op_error,
+                   bson_error_t *error)
+{
+   BSON_ASSERT_PARAM (em);
+   BSON_ASSERT (id || true);
+   BSON_ASSERT_PARAM (op_error);
+   BSON_ASSERT (error || true);
+
+   if (!id) {
+      return true;
+   }
+
+   mongoc_array_t *array = entity_map_get_bson_array (em, id, error);
+
+   if (!array) {
+      return false;
+   }
+
+   const int64_t usecs = usecs_since_epoch ();
+   const double secs = (double) usecs / 1000000.0;
+
+   bson_t *doc = bson_new ();
+   BSON_APPEND_UTF8 (doc, "error", op_error->message);
+   BSON_APPEND_DOUBLE (doc, "time", secs);
+
+   _mongoc_array_append_val (array, doc); // Transfer ownership.
+
+   return true;
+}
+
+// Expected to be set by a SIGINT (on Linux and OS X) or CTRL_BREAK_EVENT (on
+// Windows) handler when test suite is being executed by test-atlas-executor.
+bool operation_loop_terminated = false;
```

Comment:
This should probably be of type `sig_atomic_t`

Reply:
Should probably be `volatile` as well. I'll find the reference for that in the Linux Programming Interface.

Reply:
![IMG_1414](https://user-images.githubusercontent.com/13429210/228046616-ae68c5da-2b77-4422-bef0-0c0ae6ba186c.png)